Cloud9 integration for Troubleshooting (#209)
aws-samples/cfn101-workshop: content/basics/operations/troubleshooting-provisioning-errors/index.md

    2. describe both queues as [First-In-First-Out](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-sqs-queues.html#aws-sqs-queue-fifoqueue) (FIFO) queues in this example. When you describe a *source* queue and a *dead-letter* [queue](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-sqs-queues.html#aws-sqs-queue-redrive), both queues must be of the same type: either *standard* or *FIFO* (the latter is the example for this lab). Note, that the example template contains different configuration for the two `FifoQueue` property values, instead of the expected value of `true` for both. Moreover, when you describe a FIFO queue, its name must contain a `.fifo` suffix: whilst `SourceQueue` contains the `.fifo` suffix in its `QueueName`, it is configured as a standard queue (i.e., with `FifoQueue` set to `false`), thus resulting in an error.

Use the `sqs-queues.yaml` template, that contains the error mentioned earlier, to test the stack rollback pause functionality, then fix the error, and complete the stack creation:
:::::tabs{variant="container"}
::::tab{id="cloud9" label="Cloud9"}

> 1. In the **Cloud9 terminal** navigate to `code/workspace/troubleshooting-provisioing-errors`:

Reviewer: ```suggestion
1. In the **Cloud9 terminal** navigate to `code/workspace/troubleshooting-provisioning-errors`:
```

Reviewer: Typo in path

:::code{language=shell showLineNumbers=false showCopyAction=true}
cd cfn101-workshop/code/workspace/troubleshooting-provisioing-errors
:::
1. Use the AWS CLI to create the stack. The required parameter `--template-body` have been pre-filled for you.

> :::code{language=shell showLineNumbers=false showCopyAction=false}

Reviewer: ```suggestion
:::code{language=shell showLineNumbers=false showCopyAction=true}
```

aws cloudformation create-stack --stack-name cfn-workshop-troubleshooting-provisioning-errors --template-body file://sqs-queues.yaml --disable-rollback
:::
1. If the `create-stack` command was successfully sent, CloudFormation will return `StackId`.
:::code{language=shell showLineNumbers=false showCopyAction=false}
"StackId": "arn:aws:cloudformation:us-east-1:123456789012:stack/cfn-workshop-troubleshooting-provisioning-errors/739fafa0-e4d7-11ed-a000-12d9009553ff"
:::
 1. Open the **[AWS CloudFormation](https://console.aws.amazon.com/cloudformation)** console in a new tab and check if the stack status is **CREATE_FAILED**.
::::

::::tab{id="local" label="Local development"}
1. Navigate to the [AWS CloudFormation Console](https://console.aws.amazon.com/cloudformation/), and [choose a Region](https://docs.aws.amazon.com/awsconsolehelpdocs/latest/gsg/select-region.html) you wish to use.
1. From **Create stack**, choose **With new resources (standard)**.
1. From **Prepare template**, choose **Template is ready**.

> 1. From Prepare template, choose Template is ready.

Reviewer: Repeated step, please remove

1. From Template source, choose Upload a template file. Choose the `sqs-queues.yaml` template mentioned earlier, and then choose Next.
1. Specify a stack name: for example, `cfn-workshop-troubleshooting-provisioning-errors`. On the same page, accept the default value for the `QueueNamePrefix` parameter, and choose **Next**.
1. In **Stack failure options**, select **Preserve successfully provisioned resources**. Choose **Next**.
1. Choose **Create stack** in the next page.
1. Refresh the stack creation page until you see your stack in the `CREATE_FAILED` status.
::::
:::::

> The stack creation has failed, because of the error mentioned earlier. Select the name of your stack from the list (for example, `troubleshoot-provisioning-errors-workshop`): in the **Resources** tab, note that the `DeadLetterQueue` resource is in the `CREATE_COMPLETE` status, and the `SourceQueue` resource is in `CREATE_FAILED` status, along with a relevant error in the **Status reason** column.

Reviewer: ```suggestion
The stack creation has failed, because of the error mentioned earlier. Select the name of your stack from the list (for example, `cfn-workshop-troubleshooting-provisioning-errors`): in the **Resources** tab, note that the `DeadLetterQueue` resource is in the `CREATE_COMPLETE` status, and the `SourceQueue` resource is in `CREATE_FAILED` status, along with a relevant error in the **Status reason** column.
```

Your goal is to troubleshoot and fix the error in the template, and resume provisioning, to create the `SourceQueue` resource. As part of this process, preserve the state of your `DeadLetterQueue` that has been created successfully earlier. Next steps:

 With the `sqs-queues.yaml` template opened in your editor, find `SourceQueue` resource, and change `FifoQueue: false` into `FifoQueue: true`. When done, save your changes.

:::::tabs{variant="container"}
::::tab{id="cloud9" label="Cloud9"}

> 1. In the **Cloud9 terminal** navigate to `code/workspace/troubleshooting-provisioing-errors`:

Reviewer: ```suggestion
1. In the **Cloud9 terminal** navigate to `code/workspace/troubleshooting-provisioning-errors`:
```

:::code{language=shell showLineNumbers=false showCopyAction=true}
cd cfn101-workshop/code/workspace/troubleshooting-provisioing-errors
:::
1. Use the AWS CLI to update the stack. The required parameter `--template-body` have been pre-filled for you.

> :::code{language=shell showLineNumbers=false showCopyAction=false}

Reviewer: ```suggestion
:::code{language=shell showLineNumbers=false showCopyAction=true}
```

aws cloudformation update-stack --stack-name cfn-workshop-troubleshooting-provisioning-errors --template-body file://sqs-queues.yaml --disable-rollback
:::
1. If the `update-stack` command was successfully sent, CloudFormation will return `StackId`.
:::code{language=shell showLineNumbers=false showCopyAction=false}
"StackId": "arn:aws:cloudformation:us-east-1:123456789012:stack/cfn-workshop-troubleshooting-provisioning-errors/739fafa0-e4d7-11ed-a000-12d9009553ff"
:::
 1. Open the **[AWS CloudFormation](https://console.aws.amazon.com/cloudformation)** console in a new tab and check if the stack status is **UPDATE_COMPLETE**.
::::
::::tab{id="local" label="Local development"}
1. In the **Stack rollback paused** view shown in the picture earlier, choose **Update**.
1. In **Prepare template**, choose **Replace current template**, and choose to upload the template you just updated.
1. Choose **Next** in the **Parameters** page.
1. In the **Configure stack options** page, locate the **Stack failure options** section: the **Preserve successfully provisioned resources** option you chose at stack creation should still be selected. Scroll down on the page, and choose **Next**.

> 1. Next, choose **Update stack**.

Reviewer: ```suggestion
1. Next, choose **Submit**.
```

  aws cloudformation update-stack --stack-name cfn-workshop-troubleshooting-provisioning-errors --template-body file://sqs-queues.yaml --disable-rollback
  :::
1. If the `update-stack` command was successfully sent, CloudFormation will return `StackId`.
  :::code{language=shell showLineNumbers=false showCopyAction=false}
  "StackId": "arn:aws:cloudformation:us-east-1:123456789012:stack/cfn-workshop-troubleshooting-provisioning-errors/739fafa0-e4d7-11ed-a000-12d9009553ff"
  :::
1. Open the **[AWS CloudFormation](https://console.aws.amazon.com/cloudformation)** console in a new tab and check if the stack status is **UPDATE_COMPLETE**.
::::
::::tab{id="local" label="Local development"}
1. Click on the stack name, for example **cfn-workshop-troubleshooting-provisioning-errors**.
1. In the top right corner click on **Update**.
1. In **Prepare template**, choose **Replace current template**.
1. In **Template source**, choose **Upload a template file**.
1. Click on **Choose file** button and navigate to your workshop directory.
1. Select the file `sqs-queues.yaml ` and click **Next**.
1. You can leave **Configure stack options** default, click **Next**.

> 1. On the **Review <stack_name>** page, scroll down to the bottom and click on **Update stack**.

Reviewer: ```suggestion
1. On the **Review <stack_name>** page, scroll down to the bottom and click on **Submit**.
```
